src/main/kotlin/com/sourcegraph/cody/edit/FixupService.kt

```diff
@@ -0,0 +1,166 @@
+import com.intellij.openapi.diagnostic.Logger
+import com.intellij.openapi.editor.Editor
+import com.intellij.openapi.project.Project
+import com.sourcegraph.cody.agent.CodyAgentService
+import com.sourcegraph.config.ConfigUtil.isCodyEnabled
+import com.sourcegraph.utils.CodyEditorUtil
+
+/** Controller for commands that allow the LLM to edit the code directly. */
+@Service(Service.Level.PROJECT)
+class FixupService(val project: Project) : Disposable {
+  private val logger = Logger.getInstance(FixupService::class.java)
+
+  // We only use this for multiplexing task updates from the Agent to concurrent sessions.
+  // TODO: Consider doing the multiplexing in CodyAgentClient instead.
+  private var activeSessions: MutableMap<String, FixupSession> = mutableMapOf()
+
+  private var lastSelectedModel = "GPT-3.5"
+
+  // Sessions for which we have not yet received a task ID, but may receive an edit anyway.
+  private var pendingSessions: MutableSet<FixupSession> = mutableSetOf()
+
+  // The last text the user typed in without saving it, for continuity.
+  private var lastPrompt: String = ""
+
+  init {
+    // JetBrains docs say avoid heavy lifting in the constructor, so pass to another thread.
+    backgroundThread {
+      CodyAgentService.withAgent(project) { agent ->
+        ApplicationManager.getApplication().invokeLater {
+          agent.client.setOnEditTaskDidUpdate { task ->
+            val session = activeSessions[task.id]
+            if (session == null) {
+              logger.warn("No session found for task ${task.id}")
+            } else {
+              session.update(task)
+            }
+          }
+
+          agent.client.setOnEditTaskDidDelete { task ->
+            val session = activeSessions[task.id]
+            if (session == null) {
+              logger.warn("No session found for task ${task.id}")
+            } else {
+              session.taskDeleted()
+            }
+          }
+
+          agent.client.setOnWorkspaceEdit { params ->
+            for (op in params.operations) {
+              when (op.type) {
+                "create-file" -> {
+                  logger.warn("Workspace edit operation created a file: ${op.uri}")
+                }
+                "rename-file" -> {
+                  logger.warn(
+                      "Workspace edit operation renamed a file: ${op.oldUri} -> ${op.newUri}")
+                }
+                "delete-file" -> {
+                  logger.warn("Workspace edit operation deleted a file: ${op.uri}")
+                }
+                "edit-file" -> {
+                  if (op.edits == null) {
+                    logger.warn("Workspace edit operation has no edits")
+                  } else {
+                    // If there is a pending session, assume that it is the one that caused the
+                    // edit.
+                    val session: FixupSession? =
+                        if (pendingSessions.isNotEmpty()) {
```

Comment:
I'm not sure if I fully understands Fixups lifecycle.
Could you in a few sentences summarise e.g. when a Fixup is supposed to have task id and when not?

Reply:
This is a problem with the protocol that needs to be fixed; when we get editing operations, we don't know which task they are associated with (no taskId is passed).